### PR TITLE
features/SMA-25-audit-remediation

### PR DESCRIPTION
**Pull request by AmanRaj1608**
Fixes for all concern raised by Zellic team

- **3.1** The signature verification method `Secp256r1.Verify` doesn't check if the R and S !=0, which can lead to bypass the trivial signature.
- **3.2** Update the incorrect decoding of signature in `_verifySignature` method which will result in failing of transactions.
- **3.3** Google's Wycheproof have test vectors which should be run on all cryptography libs. Added the tests with the [test vector dump](https://github.com/google/wycheproof/blob/master/testvectors/ecdsa_secp256r1_sha256_test.json)
- **3.4** Remove the 1500 gas limit. We used the 1500 with estimations of EIP2165 (update precompile with gas optimisation) but some chains like BSC (will add on Aug 31, 23) don't support it and some transactions might fail with low gas limit on these chains.

### CODE REVIEW

**Comment by livingrockrises** [comment]
please add the description in PR also when you have time.

**Comment by AmanRaj1608** [comment]
TODO: 
- **3.5** Secp256r1 curve test failure, while running the whycheproof it fails for these cases 
	-  Extreme value for k and s^-1 - [Sage Script](https://sagecell.sagemath.org/?z=eJyNVFtu3DYU_RegPRDIj5Q6Bt-PAC4gaUgvoB_9CJKAlEhbgDMzljTuzBK6jG6tK-mlNHaRtgFKEJJIiOeec8-R3qHjRzQfYz-mMc5oisOpjwOKT0_jcRl79DAdTseyOFIh0R3CZ3cdeBsE_2C4H4yyeIcAayKoP00vEXnk9wMK6MPP6PKFoj9__wOdvzD0E_JnuISy8FvllcAHxMoivFIRTSeZ2OmmYY1hVrXMtu1OCCWN1rLtpCA7LFvcdYK12EnWdpZ1zFK1oxLzNlNp_RzR8TDuF3QPHKrzDbrU6JDQ8hjRfApX-Q_ntaJsidoRRy2hHadcOQ2Q0grJGs6xo0phpnaa0J1tGWuw4w1hhgNHoztqJABdViDuLOPUUmdJo5xptbXKtrxRHXbGEkkzVQZKWkaE7WzXtpJjqZlqnSBOZOr7j-gwDXH6J1tUPcV5hi2_R8e6LJ6f_8u4fxmUlbim2TWKKGM1d6w1XdNRB02mQpCyyEW7Kfolu5bG_QgPEJqnITPYqGSXysI5qHjvqryq86ldhNfzqc3zwwu8CvyW6_HfxuXxLYMD-jWOcZqXyYOK13gc_eS_xQX2y8ICur3ms8uA1SfnqpyT-gbBUw5I_Xkt_EtcvkvbyhIQbue4fF0X1fNz_Z2wFz-NPjyBtMO09jW8RaQs7nPpCmo8nLdaD5e63joDmsZ0QfP4sPfLaXpTOc5bbfiGTmGLEca99EoRQRWmnEJsDCUKe5l0jFExERKjQSWjfZLSRD9QYzSmkdFB9YxgtYJdrmChZ4EHEaXwQWsVwsA4ZIcGFoUkhkqSIvNYsUhTGpTmhCZFpZZqCEIbYDatSKpPVAWi9YAZTypqLyjTmGEAJ75nPdYG2ADdRElgwkscuE49V0DSKF0W84oT5LBOyk2eYrv9PSUVYVAemxB73ntNQj8MJuk0yCBZ35fF44YTBDBInCbTZxVx4Ay0CAMt8yQyjJmM0A8aoI-91lh5TkyKPRATmLLsy0wAatyDFXP8-u0wVPMNsreb89k5yFicFlRVj-g9mkkN13v471TTe1i8t1X27AblZtf17flS1Z_wZ3R3h6bN9DMCTLS_7hwnCEn1_7Hq1yNT_RfXj37l&lang=sage&interacts=eJyLjgUAARUAuQ==)
	-  k*G has a large x-coordinate

Recommendations:
The likely culprit is lack of consideration for the point at infinity when adding or doubling with modified Jacobians. The code comments link to the website [Point-at-Infinity](http://point-at-infinity.org/ecc/Prime_Curve_Modified_Jacobian_Coordinates.html) where the optimized algorithm used in the code is broken down. Some of the conditionals, where an early return POINT_AT_INFINITY is reached, are missing from the contract code. These checks should be implemented properly to check if the point at infinity occurs during calculations.
We also strongly reccommend to implement comprehensive test cases that go be- yond test vectors but also target edge cases in the calculations.

File: contracts/smart-account/modules/PasskeyValidationModules/Secp256r1.sol [hunk 1]
```diff
@@ -358,7 +358,7 @@ library Secp256r1 {
             mstore(add(freemem, 0x80), _exp)
             mstore(add(freemem, 0xa0), _mod)
 
-            let success := staticcall(1500, 0x5, freemem, 0xc0, freemem, 0x20)
+            let success := staticcall(not(0), 0x5, freemem, 0xc0, freemem, 0x20)
```
Comment (livingrockrises) [inline contracts/smart-account/modules/PasskeyValidationModules/Secp256r1.sol:361]: what does not(0) do?
shouldn't it be just 0

File: test/module/PasskeyFlow.module.specs.ts [hunk 2]
```diff
@@ -92,7 +92,7 @@ describe("Passkeys Registry Module:", function () {
       "nonce"
     );
     userOp1.signature = "";
-    console.log("userOp1 before", userOp1);
+    // console.log("userOp1 before", userOp1);
     // add validator module address to the signature
```
Comment (livingrockrises) [inline test/module/PasskeyFlow.module.specs.ts:96]: can remove comments if not needed

**Comment by livingrockrises** [comment]
have we added test cases mentioned by audit (some edge cases) as negative test cases?

**Review by livingrockrises — COMMENTED** [review]
reviewed.

**Comment by livingrockrises** [comment]
why there is this one json file added as part of the tests?

File: contracts/smart-account/modules/PasskeyValidationModules/Secp256r1.sol (hunk 1 above)
Comment (AmanRaj1608) [reply contracts/smart-account/modules/PasskeyValidationModules/Secp256r1.sol:361]: The not is just the bitwise NOT operation.
`not(0) - not(000....0) -> (111....11) -> max int`

The second argument represents the gas to be used for the call. So, using not(0) essentially provides the maximum amount of gas possible.

Comment (livingrockrises) [reply contracts/smart-account/modules/PasskeyValidationModules/Secp256r1.sol:361]: got it

**Comment by AmanRaj1608** [comment]
> why there is this one json file added as part of the tests?

As mentioned in audit all the cryptographic curve should pass the Wycheproof test's cases. This json is taken from - https://github.com/google/wycheproof/blob/master/testvectors/ecdsa_secp256r1_sha256_test.json

**Comment by livingrockrises** [comment]
> TODO:
> 
> * **3.5** Secp256r1 curve test failure, while running the whycheproof it fails for these cases
>   
>   * Extreme value for k and s^-1 - [Sage Script](https://sagecell.sagemath.org/?z=eJyNVFtu3DYU_RegPRDIj5Q6Bt-PAC4gaUgvoB_9CJKAlEhbgDMzljTuzBK6jG6tK-mlNHaRtgFKEJJIiOeec8-R3qHjRzQfYz-mMc5oisOpjwOKT0_jcRl79DAdTseyOFIh0R3CZ3cdeBsE_2C4H4yyeIcAayKoP00vEXnk9wMK6MPP6PKFoj9__wOdvzD0E_JnuISy8FvllcAHxMoivFIRTSeZ2OmmYY1hVrXMtu1OCCWN1rLtpCA7LFvcdYK12EnWdpZ1zFK1oxLzNlNp_RzR8TDuF3QPHKrzDbrU6JDQ8hjRfApX-Q_ntaJsidoRRy2hHadcOQ2Q0grJGs6xo0phpnaa0J1tGWuw4w1hhgNHoztqJABdViDuLOPUUmdJo5xptbXKtrxRHXbGEkkzVQZKWkaE7WzXtpJjqZlqnSBOZOr7j-gwDXH6J1tUPcV5hi2_R8e6LJ6f_8u4fxmUlbim2TWKKGM1d6w1XdNRB02mQpCyyEW7Kfolu5bG_QgPEJqnITPYqGSXysI5qHjvqryq86ldhNfzqc3zwwu8CvyW6_HfxuXxLYMD-jWOcZqXyYOK13gc_eS_xQX2y8ICur3ms8uA1SfnqpyT-gbBUw5I_Xkt_EtcvkvbyhIQbue4fF0X1fNz_Z2wFz-NPjyBtMO09jW8RaQs7nPpCmo8nLdaD5e63joDmsZ0QfP4sPfLaXpTOc5bbfiGTmGLEca99EoRQRWmnEJsDCUKe5l0jFExERKjQSWjfZLSRD9QYzSmkdFB9YxgtYJdrmChZ4EHEaXwQWsVwsA4ZIcGFoUkhkqSIvNYsUhTGpTmhCZFpZZqCEIbYDatSKpPVAWi9YAZTypqLyjTmGEAJ75nPdYG2ADdRElgwkscuE49V0DSKF0W84oT5LBOyk2eYrv9PSUVYVAemxB73ntNQj8MJuk0yCBZ35fF44YTBDBInCbTZxVx4Ay0CAMt8yQyjJmM0A8aoI-91lh5TkyKPRATmLLsy0wAatyDFXP8-u0wVPMNsreb89k5yFicFlRVj-g9mkkN13v471TTe1i8t1X27AblZtf17flS1Z_wZ3R3h6bN9DMCTLS_7hwnCEn1_7Hq1yNT_RfXj37l&lang=sage&interacts=eJyLjgUAARUAuQ==)
>   * k*G has a large x-coordinate
> 
> Recommendations: The likely culprit is lack of consideration for the point at infinity when adding or doubling with modified Jacobians. The code comments link to the website [Point-at-Infinity](http://point-at-infinity.org/ecc/Prime_Curve_Modified_Jacobian_Coordinates.html) where the optimized algorithm used in the code is broken down. Some of the conditionals, where an early return POINT_AT_INFINITY is reached, are missing from the contract code. These checks should be implemented properly to check if the point at infinity occurs during calculations. We also strongly reccommend to implement comprehensive test cases that go be- yond test vectors but also target edge cases in the calculations.

@AmanRaj1608 did you inform Kawach about this?

**Review by Aboudjem — APPROVED** [review]

**Comment by Aboudjem** [comment]
name of the PR has been changed but not the branch name

File: test/module/PasskeyFlow.module.specs.ts [hunk 3]
```diff
@@ -92,7 +92,7 @@ describe("Passkeys Registry Module:", function () {
       "nonce"
     );
     userOp1.signature = "";
-    console.log("userOp1 before", userOp1);
+    // console.log("userOp1 before", userOp1);
```
Comment (Aboudjem) [inline test/module/PasskeyFlow.module.specs.ts:95]: please remove any commented code

File: test/module/Wycheproof.specs.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import { expect } from "chai";
+import { BigNumber, Contract } from "ethers";
+import { ethers } from "hardhat";
+import testVectors from "./PasskeyModule/ecdsa_secp256r1_sha256_test.json";
+import { sha256 } from "ethereumjs-util";
+// import elliptic from "elliptic";
+// const ec = new elliptic.ec("p256");
```
Comment (Aboudjem) [inline test/module/Wycheproof.specs.ts:7]: same here

File: test/module/Wycheproof.specs.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+describe("Passkeys Registry Module:", function () {
+  // const [alice] = waffle.provider.getWallets();
```
Comment (Aboudjem) [inline test/module/Wycheproof.specs.ts:10]: +1

File: contracts/smart-account/test/TestSecp256r1.sol [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+            let z1z1 := mulmod(p3, p3, pd) // Z1Z1 = Z1^2
+            let z2z2 := mulmod(q3, q3, pd) // Z2Z2 = Z2^2
+
+            u1 := mulmod(p1, z2z2, pd) // U1 = X1*Z2Z2
+            u2 := mulmod(q1, z1z1, pd) // U2 = X2*Z1Z1
+
+            s1 := mulmod(p2, mulmod(z2z2, q3, pd), pd) // S1 = Y1*Z2*Z2Z2
+            s2 := mulmod(q2, mulmod(z1z1, p3, pd), pd) // S2 = Y2*Z1*Z1Z1
+
+            let p3q3 := addmod(p3, q3, pd)
```
Comment (ankurdubey521) [inline contracts/smart-account/test/TestSecp256r1.sol:296]: moving this to line 334 seems to fix the issue forge is having compiling the contracts

File: test/module/PasskeyFlow.module.specs.ts [hunk 7]
```diff
@@ -92,21 +90,5 @@ describe("Passkeys Registry Module:", function () {
...
-    // const handleOpsTxn = await entryPoint.handleOps(
-    //   [userOp1],
-    //   await offchainSigner.getAddress(),
-    //   {
-    //     gasLimit: 20000000,
-    //   }
-    // );
-    // await handleOpsTxn.wait();
   });
 });
```
Comment (livingrockrises) [inline test/module/PasskeyFlow.module.specs.ts:94]: "Can send a userOp with a default validation module" test case is not clear or incomplete in this file.

**Review by livingrockrises — COMMENTED** [review]
reviewed

**Review by livingrockrises — APPROVED** [review]
